### PR TITLE
Changed IndexManager and MetricManager inits.

### PR DESCRIPTION
**Pull request by jeffpierce**
This should address Issue #108.

- Bumped version to 1.3.0
- Added bootstrap option.
- If bootstrap is true, IndexManager will push the mapping for the
  cassabon index to ElasticSearch.
- Changed MetricManager's init to populate rollup paths from
  ElasticSearch if bootstrap is false (as an index should exist).

### CODE REVIEW

**Comment by jeffpierce** [comment]
@mredivo, if you could give this a code review, I'd appreciate it.

@utkarshcmu, make a build with this branch and see if that stops the goroutine explosion under the load you're throwing at it.

**Comment by utkarshcmu** [comment]
@jeffpierce I am still getting the following error under the same load which Carbon is able to handle:
```
race: limit on 8192 simultaneously alive goroutines is exceeded, dying
exit status 1
```

The error is most probably in IndexManager as the debug lines above this fatal error are from IndexManager indexing incoming paths.

**Comment by jeffpierce** [comment]
Well, damn.

@utkarshcmu:  Out of curiosity, how many stats are you sending at a time, and per batch?

**Comment by utkarshcmu** [comment]
30K metrics in batches of 300 is pushed to cassabon at a time. Each character in my metric name is approximately 100 characters.

**Comment by jeffpierce** [comment]
Awesome, thanks!

I'm about to head to bed, but I'll bang on this more tomorrow.

**Comment by utkarshcmu** [comment]
Great work !!! Gn

File: datastore/indexmanager.go [hunk 1]
```diff
@@ -87,6 +91,99 @@ func (im *IndexManager) run() {
...
+
+		for _, hit := range esResp.Hits.Hits {
+			pathList = append(pathList, hit.Source.Path)
+		}
+	} else {
+		logging.Statsd.Client.Inc("indexmgr.es.err.get", 1, 1.0)
+		config.G.Log.System.LogError("Error querying ES.")
+	}
+
+	return pathList
```
Comment (mredivo) [inline datastore/indexmanager.go]: You might want to issue a debug log entry indicating the number of paths retrieved, so you can see if that matches expectations.

**Comment by mredivo** [comment]
As the goroutine pool is still exhausted after this change, this isn't the whole fix, but it looks like this PR should solve the "thundering herd" of path indexing on restarts.

LGTM

**Comment by utkarshcmu** [comment]
+1 for this PR to get merged anyway.

**Comment by jeffpierce** [comment]
Okay.  Added the debug log that @mredivo requested.

@utkarshcmu:  I put a queuing system that should limit the number of goroutines uses during storage.  Try a build with the newest changes and see if that handles it.  If it's slow, we can always ramp the number of workers on that queue up (or just make it configurable via CLI flag and yaml config).

**Comment by utkarshcmu** [comment]
Tested. Able to take load now. Not dying with the last error.

MetricsManager is queueing write batches, which is increasing in case of 30K events at a time in 10 seconds interval. Looks like I need more Cassabon nodes. :)

But yeah queueing solution worked to get that dying error resolved. Lets merge this to master. 👍

**Comment by jeffpierce** [comment]
@utkarshcmu I'll address that comment over in the issue thread -- there are some tweaks you can give the Cassandra settings in the cassabon.yaml file to overcome that.

Merging.

File: datastore/indexmanager.go [hunk 2]
```diff
@@ -12,6 +12,8 @@ import (
 
 	"github.com/jeffpierce/cassabon/config"
 	"github.com/jeffpierce/cassabon/logging"
+
+	"github.com/otium/queue"
```
Comment (mredivo) [inline datastore/indexmanager.go]: Go convention states that this import should be above our own package imports, still on its own line.

File: datastore/indexmanager.go [hunk 3]
```diff
@@ -56,14 +58,21 @@ type ERQuery struct {
...
 func (im *IndexManager) Init(bootstrap bool) {
 	// If bootstrap is true, initialize mapping in ElasticSearch
 	if bootstrap {
 		im.initMapping()
 	}
+
+	// Initialize index worker queue.
+	im.IndexQueue = queue.NewQueue(func(metricPath interface{}) {
+		path, _ := metricPath.(string)
+		im.index(path)
```
Comment (mredivo) [inline datastore/indexmanager.go]: Consider
```
if path, ok := metricPath.(string); ok {
    im.index(path)
}
```

File: datastore/indexmanager.go (hunk 2 above)
Comment (mredivo) [inline datastore/indexmanager.go]: Note to self: "github.com/otium/queue" needs to be vendored in https://github.com/jeffpierce/cassabon/pull/110

**Pull request merged**